mu4e/mu4e-actions.el

```diff
@@ -217,35 +217,42 @@ store your org-contacts."
 
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defvar mu4e--patch-directory-history nil
+  "History of directories we have applied patches to.")
+
+;; This essentially works around the fact that read-directory-name
+;; can't have custom history.
+(defun mu4e--read-patch-directory (&optional prompt)
+  "Read a `PROMPT'ed directory name via `completing-read' with history."
+  (unless prompt
+    (setq prompt "Target directory:"))
+  (file-truename
+   (completing-read prompt 'read-file-name-internal #'file-directory-p
+                    nil nil 'mu4e--patch-directory-history)))
+
 (defun mu4e-action-git-apply-patch (msg)
-  "Apply the git [patch] message."
-  (let ((path (ido-read-directory-name "Target directory: "
-		(car ido-work-directory-list)
-		"~/" t)))
-    (setf ido-work-directory-list
-      (cons path (delete path ido-work-directory-list)))
+  "Apply `MSG' as a git patch."
+  (let ((path (mu4e--read-patch-directory "Target directory: ")))
     (shell-command
-      (format "cd %s; git apply %s"
-	path
-	(mu4e-message-field msg :path)))))
+     (format "cd %s; git apply %s"
+             path
+             (shell-quote-argument (mu4e-message-field msg :path))))))
```

Comment:
when we're changing things here anyway, `path` should be escaped here as well, and we should let-bind `default-directory` rather than `cd` in the shell-command.